File: kinotic-frontend/packages/common/src/components/deploy/ProjectDeployStores.ts
import type { ProjectArtifacts } from '@kinotic-ai/management-api'
import type { JobTaskNode } from '../grind/JobTaskNode'

/** Mirrors DeployTarget on the server: what a deployment run's first task decided. */
export interface DeployTarget {
  nodeId: string
  hostDir: string
  syncWorkloadId: string
  uiPublishWorkloadId: string
}

/**
 * The job scope names a project deployment run stores its outcomes under, mirroring
 * ProjectDeployStores on the server, and what they mean for the job page: a task node
 * carrying one of them is a task that ran a workload, which is how the page attaches that
 * workload's log to the task's row.
 */
export default class ProjectDeployStores {

  public static readonly ARTIFACTS = 'artifacts'
  public static readonly DEPLOY_TARGET = 'deployTarget'
  public static readonly SYNC_WORKLOAD_ID = 'syncWorkloadId'
  public static readonly UI_DEPLOYMENTS = 'uiDeployments'

  /** The artifacts the task bound into the run, or null while the task has not completed. */
  public static artifactsOf(node: JobTaskNode): ProjectArtifacts | null {
    let ret: ProjectArtifacts | null = null
    if (ProjectDeployStores.hasArtifacts(node) && node.storedValue !== null && node.storedValue !== undefined) {
      ret = node.storedValue as ProjectArtifacts
    }
    return ret
  }

  /** Whether the task's row lists the artifacts the deployed commit contains. */
  public static hasArtifacts(node: JobTaskNode): boolean {
    return node.storedName === ProjectDeployStores.ARTIFACTS
  }

  /** Whether the task's row has a detail pane: the artifacts it bound, or the workload log it ran. */
  public static hasDetail(node: JobTaskNode): boolean {
    return ProjectDeployStores.hasArtifacts(node) || ProjectDeployStores.hasWorkloadLog(node)
  }

  /**
   * Whether the task's row carries a workload log: the sync task's is the run's build log, the
   * publish task's is the upload log.
   */
  public static hasWorkloadLog(node: JobTaskNode): boolean {
    return node.storedName === ProjectDeployStores.SYNC_WORKLOAD_ID
      || node.storedName === ProjectDeployStores.UI_DEPLOYMENTS
  }

  /**
   * The workload whose log belongs to the task, or null while it is not yet known. Both
   * workloads are named by the resolved deploy target before their tasks run; the sync task
   * also names its own once it completed.
   */
  public static workloadLogOf(node: JobTaskNode, root: JobTaskNode | null): string | null {
    let ret: string | null = null
    const target = root?.children
      .find(child => child.storedName === ProjectDeployStores.DEPLOY_TARGET)
      ?.storedValue as DeployTarget | undefined
    if (node.storedName === ProjectDeployStores.SYNC_WORKLOAD_ID) {
      ret = typeof node.storedValue === 'string' ? node.storedValue : target?.syncWorkloadId ?? null
    } else if (node.storedName === ProjectDeployStores.UI_DEPLOYMENTS) {
      ret = target?.uiPublishWorkloadId ?? null
    }
    return ret
  }
}

File: kinotic-system-api/src/main/java/org/kinotic/system/api/model/deployment/ProjectDeployStores.java
package org.kinotic.system.api.model.deployment;

/**
 * The job scope names a project deployment run stores its outcomes under. They reach a
 * watcher of the run on its {@code TaskCompletedEvent}s and outlive it on its
 * {@code TaskRecord}s, which is how the console finds the workloads whose logs belong to a
 * run.
 */
public final class ProjectDeployStores {

    /** The resolved {@link DeployTarget}, including the id of the run's sync workload. */
    public static final String DEPLOY_TARGET = "deployTarget";

    /**
     * The {@link org.kinotic.management.api.model.ProjectArtifacts} of the deployed commit,
     * as the sync workload found them in the checkout and the run bound them.
     */
    public static final String ARTIFACTS = "artifacts";

    /**
     * The id of the run's sync workload, stored by the task that ran it. Known before that
     * task completes through {@link DeployTarget#syncWorkloadId()}, which is what lets a
     * watcher follow the workload's logs while the task runs.
     */
    public static final String SYNC_WORKLOAD_ID = "syncWorkloadId";

    /**
     * The {@link MicroserviceDeployments} the run left the project with, one per microservice
     * of the project, ensured or orphaned.
     */
    public static final String MICROSERVICE_DEPLOYMENTS = "microserviceDeployments";

    /**
     * The {@link UiDeployments} the run left the project with, one per UI of the project,
     * published or orphaned. Known before the task completes through
     * {@link DeployTarget#uiPublishWorkloadId()}, which is what lets a watcher follow the
     * publish workload's logs while the task runs.
     */
    public static final String UI_DEPLOYMENTS = "uiDeployments";

    private ProjectDeployStores() {
    }
}
